### ISSUE TITLE
Type changes in 3.23 

### ISSUE THREAD

**Issue by juliusmarminge** [colinhacks/zod#3431]
Hi Colin,

@t3-oss/t3-env has started failing in zod@3.23: https://github.com/t3-oss/t3-env/issues/222

We're not using any internal types which made me think you've unintentionally broken semver for this release?

See the source, we're basically just using ZodType: 
- https://github.com/t3-oss/t3-env/blob/main/packages/core/src/index.ts
- https://github.com/t3-oss/t3-env/blob/main/packages/nextjs/src/index.ts

We're also not really using the ZodType generics, so I don't think that change is relevant here. We're only using ZodType as a constraint

**Comment by colinhacks** [comment]
Sorry about this. It's likely related to the changes to `ZodType` mentioned in the changelog, though I also am surprised that it affected your package given how it uses `ZodType`. I couldn't immediately replicate this in the t3-env repo but based on https://github.com/t3-oss/t3-env/issues/222 it seems like you've found a fix? 

For others reading this here's the change: 

```diff
- class ZodType<Output = any, Def extends ZodTypeDef = ZodTypeDef, Input = Output> {}
+ class ZodType<Output = unknown, Def extends ZodTypeDef = ZodTypeDef, Input = unknown> {}
```

This fixes what I consider to be a bug in `ZodType` that has caused a lot of headache for many many people. This is the last 3.x release of Zod and I wanted to rectify this before starting on v4, since many projects will likely stay on 3.x for some time.

**Comment by masterkain** [comment]
```
#15 20.73 ./lib/actions/projectActions.ts:137:47
#15 20.73 Type error: Argument of type 'Zod.ZodIssue[]' is not assignable to parameter of type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodIssue[]'.
#15 20.73   Type 'Zod.ZodIssue' is not assignable to type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodIssue'.
#15 20.73     Type 'ZodInvalidUnionIssue & { fatal?: boolean | undefined; message: string; }' is not assignable to type 'ZodIssue'.
#15 20.73       Type 'Zod.ZodInvalidUnionIssue & { fatal?: boolean | undefined; message: string; }' is not assignable to type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodInvalidUnionIssue & { fatal?: boolean | undefined; message: string; }'.
#15 20.73         Type 'ZodInvalidUnionIssue & { fatal?: boolean | undefined; message: string; }' is not assignable to type 'ZodInvalidUnionIssue'.
#15 20.73           Types of property 'unionErrors' are incompatible.
#15 20.73             Type 'Zod.ZodError<any>[]' is not assignable to type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodError<any>[]'.
#15 20.73               Type 'Zod.ZodError<any>' is not assignable to type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodError<any>'.
#15 20.73                 Types of property 'issues' are incompatible.
#15 20.73                   Type 'Zod.ZodIssue[]' is not assignable to type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodIssue[]'.
#15 20.73                     Type 'Zod.ZodIssue' is not assignable to type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodIssue'.
#15 20.73                       Type 'ZodInvalidArgumentsIssue & { fatal?: boolean | undefined; message: string; }' is not assignable to type 'ZodIssue'.
#15 20.73 
#15 20.73   135 |   const parsedData = formDataSchema.safeParse(formData);
#15 20.73   136 |   if (!parsedData.success) {
#15 20.73 > 137 |     const errorMessage = generateErrorMessage(parsedData.error.issues);
#15 20.73       |                                               ^
#15 20.73   138 |     revalidateTag(`project_${projectId}`);
#15 20.73   139 |     return { project_id: projectId, error: errorMessage };
#15 20.73   140 |   }
```

I started seeing this in an automated build that proposes an upgrade to v3.23.0 from v3.22.5, I'll check it later but maybe it has something to do with this?

**Comment by paradoxloop** [comment]
@colinhacks this update has also broken typechecking on [zodResolver](https://github.com/react-hook-form/resolvers) of react-hooks-form in our projects.

Will do a deeper dive later if required but wanted to flag early.

`src/pages/projects/[projectId]/export/index.tsx:58:27 - error TS2345: Argument of type 'ZodObject<{ projectId: ZodString; bySearchFileStructure: ZodBoolean; searches: ZodObject<{ asicExtracts: ZodBoolean; asicRelational: ZodBoolean; trademarkSearch: ZodBoolean; insolvencySearch: ZodBoolean; whoisSearch: ZodBoolean; ppsrSearch: ZodBoolean; }, "strip", ZodTypeAny, { ...; }, { ...; }>; }, "strip", ZodType...' is not assignable to parameter of type 'ZodType<any, any, any>'.
  Types of property '_parse' are incompatible.
    Type '(input: ParseInput) => ParseReturnType<{ projectId: string; searches: { ppsrSearch: boolean; insolvencySearch: boolean; whoisSearch: boolean; asicExtracts: boolean; asicRelational: boolean; trademarkSearch: boolean; }; bySearchFileStructure: boolean; }>' is not assignable to type '(input: ParseInput) => ParseReturnType<any>'.
      Types of parameters 'input' and 'input' are incompatible.
        Type 'import("/Users/nico/Documents/projects/ddloop-dashboard/node_modules/.pnpm/zod@3.22.4/node_modules/zod/lib/helpers/parseUtil").ParseInput' is not assignable to type 'Zod.ParseInput'.
          The types of 'parent.common.issues' are incompatible between these types.
            Type 'import("/Users/nico/Documents/projects/ddloop-dashboard/node_modules/.pnpm/zod@3.22.4/node_modules/zod/lib/ZodError").ZodIssue[]' is not assignable to type 'Zod.ZodIssue[]'.
              Type 'import("/Users/nico/Documents/projects/ddloop-dashboard/node_modules/.pnpm/zod@3.22.4/node_modules/zod/lib/ZodError").ZodIssue' is not assignable to type 'Zod.ZodIssue'.
                Type 'ZodInvalidUnionIssue & { fatal?: boolean | undefined; message: string; }' is not assignable to type 'ZodIssue'.
                  Type 'import("/Users/nico/Documents/projects/ddloop-dashboard/node_modules/.pnpm/zod@3.22.4/node_modules/zod/lib/ZodError").ZodInvalidUnionIssue & { fatal?: boolean | undefined; message: string; }' is not assignable to type 'Zod.ZodInvalidUnionIssue & { fatal?: boolean | undefined; message: string; }'.
                    Type 'ZodInvalidUnionIssue & { fatal?: boolean | undefined; message: string; }' is not assignable to type 'ZodInvalidUnionIssue'.
                      Types of property 'unionErrors' are incompatible.
                        Type 'import("/Users/nico/Documents/projects/ddloop-dashboard/node_modules/.pnpm/zod@3.22.4/node_modules/zod/lib/ZodError").ZodError<any>[]' is not assignable to type 'Zod.ZodError<any>[]'.
                          Type 'import("/Users/nico/Documents/projects/ddloop-dashboard/node_modules/.pnpm/zod@3.22.4/node_modules/zod/lib/ZodError").ZodError<any>' is not assignable to type 'Zod.ZodError<any>'.
                            Types of property 'issues' are incompatible.
                              Type 'import("/Users/nico/Documents/projects/ddloop-dashboard/node_modules/.pnpm/zod@3.22.4/node_modules/zod/lib/ZodError").ZodIssue[]' is not assignable to type 'Zod.ZodIssue[]'.
                                Type 'import("/Users/nico/Documents/projects/ddloop-dashboard/node_modules/.pnpm/zod@3.22.4/node_modules/zod/lib/ZodError").ZodIssue' is not assignable to type 'Zod.ZodIssue'.
                                  Type 'ZodInvalidUnionIssue & { fatal?: boolean | undefined; message: string; }' is not assignable to type 'ZodIssue'.

58     resolver: zodResolver(ZipExportFormValuesSchema),`
<img width="1642" alt="Screenshot 2024-04-22 at 12 23 19" src="https://github.com/colinhacks/zod/assets/709833/10d1b874-07e5-468d-b850-2953493f64bd">

**Comment by colinhacks** [comment]
@masterkain That's a symptom of having two different versions of Zod installed in your project. The two `ZodError` classes aren't assignable to each other.

@paradoxloop I'd appreciate some kind of reproduction here, or some guidance on what you're seeing.

**Comment by jthrilly** [comment]
I can't provide a full repro for the react-hooks-form issue, but I can share this helper hook that I believe is widely used (including in my projects) and is now errors due to this:

```
import type { z } from 'zod';
import { useForm, type UseFormProps } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';

/**
 * Reusable hook for zod + react-hook-form
 */
export default function useZodForm<TSchema extends z.ZodType>(
  props: Omit<UseFormProps<TSchema['_input']>, 'resolver'> & {
    schema: TSchema;
  },
) {
  const form = useForm<TSchema['_input']>({
    ...props,
    resolver: zodResolver(props.schema, undefined),
  });

  return form;
}
```

Error:

`Type 'TSchema["_input"]' does not satisfy the constraint 'FieldValues'`

(FieldValues is typed as `Record<string, any>`).

It does indeed seem related to the ZodType change, because you can do `TSchema extends z.ZodType<any, z.ZodTypeDef, any>` to "get rid" of the error (albeit without proper input and output parsing).

I can create a new issue if needed. LMK.

**Comment by juliusmarminge** [comment]
Here is an updated version of the hook:

https://github.com/t3-oss/create-t3-turbo/blob/2a161360bfbd39053e518c9580e515c37dc5e77d/packages/ui/src/form.tsx#L25

**Comment by colinhacks** [comment]
Okay, given that this `useZodForm` hook is widely used, I think changing `any` to `unknown` is now too widely disruptive. I've reverted in 3.23.1.

**Comment by colinhacks** [comment]
Upgrade to Zod 3.23.3 to resolve both the t3-env issue (see [details](https://github.com/t3-oss/t3-env/issues/222#issuecomment-2071070276)) and React Hook Form issues.

**Comment by MaximeBernard** [comment]
@colinhacks We loved the change! Any timeline where this comes back?

**Comment by ryanleecode** [comment]
This drove me nuts. thanks for the fix

**Comment by charIeszhao** [comment]
> ```
> #15 20.73 ./lib/actions/projectActions.ts:137:47
> #15 20.73 Type error: Argument of type 'Zod.ZodIssue[]' is not assignable to parameter of type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodIssue[]'.
> #15 20.73   Type 'Zod.ZodIssue' is not assignable to type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodIssue'.
> #15 20.73     Type 'ZodInvalidUnionIssue & { fatal?: boolean | undefined; message: string; }' is not assignable to type 'ZodIssue'.
> #15 20.73       Type 'Zod.ZodInvalidUnionIssue & { fatal?: boolean | undefined; message: string; }' is not assignable to type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodInvalidUnionIssue & { fatal?: boolean | undefined; message: string; }'.
> #15 20.73         Type 'ZodInvalidUnionIssue & { fatal?: boolean | undefined; message: string; }' is not assignable to type 'ZodInvalidUnionIssue'.
> #15 20.73           Types of property 'unionErrors' are incompatible.
> #15 20.73             Type 'Zod.ZodError<any>[]' is not assignable to type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodError<any>[]'.
> #15 20.73               Type 'Zod.ZodError<any>' is not assignable to type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodError<any>'.
> #15 20.73                 Types of property 'issues' are incompatible.
> #15 20.73                   Type 'Zod.ZodIssue[]' is not assignable to type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodIssue[]'.
> #15 20.73                     Type 'Zod.ZodIssue' is not assignable to type 'import("/app/node_modules/zod-error/node_modules/zod/lib/ZodError").ZodIssue'.
> #15 20.73                       Type 'ZodInvalidArgumentsIssue & { fatal?: boolean | undefined; message: string; }' is not assignable to type 'ZodIssue'.
> #15 20.73 
> #15 20.73   135 |   const parsedData = formDataSchema.safeParse(formData);
> #15 20.73   136 |   if (!parsedData.success) {
> #15 20.73 > 137 |     const errorMessage = generateErrorMessage(parsedData.error.issues);
> #15 20.73       |                                               ^
> #15 20.73   138 |     revalidateTag(`project_${projectId}`);
> #15 20.73   139 |     return { project_id: projectId, error: errorMessage };
> #15 20.73   140 |   }
> ```
> 
> https://github.com/icoretech/airbroke/blob/3f8b7658c03ad2d03c73e6bcbc12633fc7d45135/lib/actions/projectActions.ts#L134-L140
> 
> I started seeing this in an automated build that proposes an upgrade to v3.23.0 from v3.22.5, I'll check it later but maybe it has something to do with this?

Had this error today in my project after upgrading the dependencies, but then I realized it was caused by having two different versions of zod in the node_modules. Removing the different version fixed the issue for me.

**Comment by asontha** [comment]
+1 to the above, was migrating from 3.22.4 to 3.23.8 and needed to make sure all microservices using zod were on the same version.

**Comment by rgalaxy** [comment]
I am using 
```
zod: 3.25.76
react-hook-form: 7.56.0
@hookform/resolvers: 5.2.2
```

and still got this error

```
No overload matches this call.
  Overload 1 of 4, '(schema: Zod3Type<any, any>, schemaOptions?: ParseParams | undefined, resolverOptions?: NonRawResolverOptions | undefined): Resolver<...>', gave the following error.
    Argument of type 'ZodType<any, ZodTypeDef, any>' is not assignable to parameter of type 'Zod3Type<any, any>'.
      Types of property '_def' are incompatible.
        Property 'typeName' is missing in type 'ZodTypeDef' but required in type '{ typeName: string; }'.
  Overload 2 of 4, '(schema: $ZodType<unknown, FieldValues, $ZodTypeInternals<unknown, FieldValues>>, schemaOptions?: ParseContext<$ZodIssue> | undefined, resolverOptions?: NonRawResolverOptions | undefined): Resolver<...>', gave the following error.
    Argument of type 'ZodType<any, ZodTypeDef, any>' is not assignable to parameter of type '$ZodType<unknown, FieldValues, $ZodTypeInternals<unknown, FieldValues>>'.
      Property '_zod' is missing in type 'ZodType<any, ZodTypeDef, any>' but required in type '$ZodType<unknown, FieldValues, $ZodTypeInternals<unknown, FieldValues>>'.ts(2769)
zod.d.ts(16, 9): 'typeName' is declared here.
schemas.d.cts(84, 5): '_zod' is declared here.
```

I don't understand what needs to be updated (I think I made sure all the required package updated to its latest version).
